march_hardware/test/TestEncoder.cpp

```diff
@@ -1,29 +1,116 @@
-using ::testing::Return;
-
-class TestEncoder : public ::testing::Test
+class TestEncoder : public testing::Test
 {
 protected:
-  const float angle = 42;
+  const int32_t lower_limit = 2053;
+  const int32_t upper_limit = 45617;
+  march::Encoder encoder =
+      march::Encoder(17, lower_limit, upper_limit, -0.34906585, 1.745329252, -0.296705973, 1.692969374);
 };
 
-TEST_F(TestEncoder, Encoder)
+class TestEncoderParameterizedLimits : public TestEncoder, public testing::WithParamInterface<std::tuple<int32_t, bool>>
 {
-  MockEncoder mockEncoder;
+};
 
-  EXPECT_CALL(mockEncoder, getAngleDeg()).Times(AtLeast(1));
-  ON_CALL(mockEncoder, getAngleDeg()).WillByDefault(Return(angle));
+class TestEncoderParameterizedSoftLimits : public TestEncoder,
+                                           public testing::WithParamInterface<std::tuple<int32_t, bool>>
+{
+};
 
-  ASSERT_EQ(angle, mockEncoder.getAngleDeg());
+class TestEncoderParameterizedValidTarget : public TestEncoder,
+                                            public testing::WithParamInterface<std::tuple<int32_t, int32_t, bool>>
+{
+};
+
+TEST_F(TestEncoder, CorrectLimits)
+{
+  ASSERT_EQ(this->encoder.getLowerHardLimitIU(), this->lower_limit);
+  ASSERT_EQ(this->encoder.getUpperHardLimitIU(), this->upper_limit);
+}
+
+TEST_F(TestEncoder, CorrectSoftLimits)
+{
+  ASSERT_EQ(this->encoder.getLowerSoftLimitIU(), 3144);
+  ASSERT_EQ(this->encoder.getUpperSoftLimitIU(), 44650);
+}
+
+TEST(TestEncoderConstructor, ResolutionBelowRange)
+{
+  ASSERT_THROW(march::Encoder(0, 0, 0, 0, 0, 0, 0), march::error::HardwareException);
+}
+
+TEST(TestEncoderConstructor, ResolutionAboveRange)
+{
+  ASSERT_THROW(march::Encoder(50, 0, 0, 0, 0, 0, 0), march::error::HardwareException);
+}
+
+TEST(TestEncoderConstructor, DifferentIuPerRad)
+{
+  ASSERT_THROW(march::Encoder(12, 2, 1, 1.0, 2.0, 1.0, 2.0), march::error::HardwareException);
+}
+
+TEST(TestEncoderConstructor, InvalidSoftLimits)
+{
+  ASSERT_THROW(march::Encoder(17, 2053, 45617, -0.34906585, 1.745329252, 0.4, 0.1), march::error::HardwareException);
+}
+
+TEST(TestEncoderConstructor, InvalidLowerSoftLimit)
+{
+  ASSERT_THROW(march::Encoder(17, 2053, 45617, -0.34906585, 1.745329252, -0.4, 1), march::error::HardwareException);
+}
+
+TEST(TestEncoderConstructor, InvalidUpperSoftLimit)
+{
+  ASSERT_THROW(march::Encoder(17, 2053, 45617, -0.34906585, 1.745329252, -0.3, 2.0), march::error::HardwareException);
+}
+
+INSTANTIATE_TEST_CASE_P(ParameterizedLimits, TestEncoderParameterizedLimits,
+                        testing::Values(std::make_tuple(2052, false), std::make_tuple(2053, false),
+                                        std::make_tuple(2054, true), std::make_tuple(45616, true),
+                                        std::make_tuple(45617, false), std::make_tuple(45618, false),
+                                        std::make_tuple(3000, true), std::make_tuple(0, false),
+                                        std::make_tuple(60001, false)));
```

Comment:
This is some nice testing :). Maybe also a test for a negative value? (which would be in range if positive)